File: algorithms/python/dijkstra.py
import sys
import time
import heapq

INF = float('inf')

def dijkstra(graph, V, src):
    dist = [INF] * V
    dist[src] = 0
    heap = [(0, src)]

    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]:
            continue
        for to, w in graph[u]:
            nd = d + w
            if nd < dist[to]:
                dist[to] = nd
                heapq.heappush(heap, (nd, to))
    return dist

def main():
    data = sys.stdin.read().split()
    idx  = 0
    V    = int(data[idx]); idx += 1
    E    = int(data[idx]); idx += 1
    src  = int(data[idx]); idx += 1

    graph = [[] for _ in range(V)]
    for _ in range(E):
        u = int(data[idx]); idx += 1
        t = int(data[idx]); idx += 1
        w = int(data[idx]); idx += 1
        graph[u].append((t, w))

    start = time.perf_counter()
    dist  = dijkstra(graph, V, src)
    ms    = (time.perf_counter() - start) * 1000

    sys.stdout.write(" ".join(str(-1 if d == INF else d) for d in dist) + "\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()

File: algorithms/julia/dijkstra.jl
using DataStructures

const INF = typemax(Int64) ÷ 2

function dijkstra(graph::Vector{Vector{Tuple{Int,Int}}}, v::Int, src::Int)
    dist = fill(INF, v)
    dist[src+1] = 0
    pq = PriorityQueue{Int,Int}()
    enqueue!(pq, src+1, 0)

    while !isempty(pq)
        u, d = dequeue_pair!(pq)
        d > dist[u] && continue
        for (to, w) in graph[u]
            nd = d + w
            if nd < dist[to]
                dist[to] = nd
                pq[to] = nd
            end
        end
    end
    return dist
end

function main()
    input = split(read(stdin, String))
    idx = 1
    V   = parse(Int, input[idx]); idx += 1
    E   = parse(Int, input[idx]); idx += 1
    src = parse(Int, input[idx]); idx += 1

    graph = [Vector{Tuple{Int,Int}}() for _ in 1:V]
    for _ in 1:E
        u = parse(Int, input[idx])+1; idx += 1
        t = parse(Int, input[idx])+1; idx += 1
        w = parse(Int, input[idx]);   idx += 1
        push!(graph[u], (t, w))
    end

    # JIT warmup
    g2 = [[(2,1)], []]
    dijkstra(g2, 2, 0)

    start = time_ns()
    dist  = dijkstra(graph, V, src)
    ms    = (time_ns() - start) / 1e6

    println(join(map(d -> d == INF ? -1 : d, dist), " "))
    println(stderr, ms)
end

main()